proposal/new-js-importer.d.ts

```diff
@@ -0,0 +1,322 @@
+ * for it to complete and uses its value as the return value. If a `Promise`
+ * emits an error, it rethrows that error.
+ */
+interface AsyncFileImporter {
+  findFileUrl(
+    url: string,
+    options: {fromImport: boolean}
+  ): Promise<FileImporterResult | null> | FileImporterResult | null;
+
+  canonicalize?: never;
+}
+
+export interface FileImporterResult {
+  /** The partially-resolved `file:` URL of a file on disk. */
+  url: URL;
+
+  /**
+   * A browser-accessible URL indicating the resolved location of the imported
+   * stylesheet.
+   *
+   * The implementation must use this URL in source maps to refer to source
+   * spans in `css`.
+   */
+  sourceMapUrl?: URL;
+}
+
+/**
+ * This interface represents an [importer]. When the importer is invoked with a
+ * string `string`:
+ *
+ * [importer]: ../spec/modules.md#importer
+ *
+ * * Let `fromImport` be `true` if the importer is being run for an `@import`
+ *   and `false` otherwise.
+ *
+ * * Let `url` be the result of calling `canonicalize` with `url` and
+ *   `fromImport`.
+ *
+ * * If `url` is null, return null.
+ *
+ * * Let `result` be the result of calling `load` with `url`.
+ *
+ * * If `result` is null, return null.
+ *
+ * * If `result.syntax` is `"scss"`, `"css"`:
+ *
+ *   * Let `syntax` be `result.syntax`.
+ *
+ * * Otherwise, if `result.syntax` is `"sass"`:
+ *
+ *   * Let `syntax` be "indented".
+ *
+ * * Otherwise, throw an error.
+ *
+ * * Return `result.css`, `syntax`, and `url`.
+ *
+ * [resolving `url`]: ../spec/modules.md#resolving-a-file-url
+ */
+interface SyncImporter {
+  canonicalize(url: string, options: {fromImport: boolean}): URL | null;
+  load(canonicalUrl: URL): ImporterResult | null;
+  findFileUrl?: never;
+}
+
+/**
+ * This interface represents an [importer]. It has the same beavior as
+ * `SyncImporter`, except that if `canonicalize` or `load` return a `Promise` it
+ * waits for it to complete and uses its value as the return value. If a
+ * `Promise` emits an error, it rethrows that error.
+ */
+interface AsyncImporter {
+  canonicalize(
+    url: string,
+    options: {fromImport: boolean}
+  ): Promise<URL | null> | URL | null;
+
+  load(
+    canonicalUrl: URL
+  ): Promise<ImporterResult | null> | ImporterResult | null;
+
+  findFileUrl?: never;
+}
+
+export interface ImporterResult {
+  /** The contents of stylesheet loaded by an importer. */
+  css: string;
+
+  /** The syntax to use to parse `css`. */
+  syntax: Syntax;
+
+  /**
+   * A browser-accessible URL indicating the resolved location of the imported
+   * stylesheet.
+   *
+   * The implementation must use this URL in source maps to refer to source
+   * spans in `css`.
+   */
+  sourceMapUrl?: URL;
+}
+
+/**
+ * > This type is exported so that TypeScript users can explicitly write a class
+ * > that `implements FileImporter`.
+ */
+export type FileImporter<sync extends 'sync' | 'async'> = sync extends 'async'
```

Comment:
wdyt about exporting `FileImporter` and `AsyncFileImporter` directly instead?

Reply:
I think if we're going to expose the generic type anywhere, we should use it consistently everywhere.